posthog/api/dashboard.py

```diff
@@ -216,15 +219,34 @@ def update(self, instance: Dashboard, validated_data: Dict, *args: Any, **kwargs
 
         validated_data.pop("use_template", None)  # Remove attribute if present
 
+        being_undeleted = instance.deleted and "deleted" in validated_data and not validated_data["deleted"]
+        if being_undeleted:
+            DashboardTile.objects.filter(dashboard__id=instance.id).update(deleted=False)
+            insights_to_undelete = []
+            for tile in DashboardTile.objects.filter(dashboard__id=instance.id):
+                if tile.insight and tile.insight.deleted:
+                    tile.insight.deleted = False
+                    insights_to_undelete.append(tile.insight)
+            Insight.objects.bulk_update(insights_to_undelete, ["deleted"])
```

Comment:
There is still a surprise waiting for users here: _All_ tiles that were previously on the dashboard are restored, so "Undo" doesn't bring the dashboard back as it was, it brings it back with the ghosts of its history. I think this can be fine for now, but it is a bug, so will probably get reported eventually. I don't have a simple idea for solving it.
One idea I have is replacing `deleted` on dashboards, tiles, and insights with `deleted_at`, which would allow us to coordinate delete/undo across all those related models, but then this PR would balloon completely.

Reply:
I know this is stale, but just wanted to push a thought - given that you're soft deleting insights, maybe a nightly/weekly task to permanently delete soft-deleted insights would at least limit the amount / staleness of the ghosts in question.

In any case, it's not a perfect solution to this problem, but it at least seems like a simple / easy / Pareto solution to it.